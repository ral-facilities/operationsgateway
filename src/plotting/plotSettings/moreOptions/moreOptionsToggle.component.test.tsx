import type { RenderResult } from '@testing-library/react';
import { render, screen } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { SelectedPlotChannel } from '../../../app.types';
import { testPlotDatasets } from '../../../testUtils';
import type { MoreOptionsProps } from './moreOptionsBox.component';
import MoreOptionsToggle from './moreOptionsToggle.component';

vi.mock('./moreOptionsBox.component', () => {
  return {
    default: (props: MoreOptionsProps) => (
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      <mock-moreOptionsBox data-testid="mock-moreOptionsBox">
        {Object.entries(props).map(
          ([propName, propValue]) =>
            `${propName}=${JSON.stringify(propValue, null, 2)}\n`
        )}
        {/* eslint-disable-next-line @typescript-eslint/ban-ts-comment */}
        {/* @ts-ignore */}
      </mock-moreOptionsBox>
    ),
  };
});

describe('MoreOptionsToggle', () => {
  let props: MoreOptionsProps;
  const changeSelectedPlotChannels = vi.fn();
  let user: UserEvent;

  const createView = (): RenderResult => {
    return render(<MoreOptionsToggle {...props} />);
  };

  const testSelectedPlotChannels: SelectedPlotChannel[] = testPlotDatasets.map(
    (dataset, i) => ({
      name: dataset.name,
      units: 'mm',
      options: {
        visible: true,
        colour: `colour-${i.toString()}`,
        lineStyle: 'solid',
        yAxis: 'left',
      },
    })
  );

  beforeEach(() => {
    props = {
      channel: testSelectedPlotChannels[0],
      selectedPlotChannels: testSelectedPlotChannels,
      changeSelectedPlotChannels,
    };

    user = userEvent.setup({ delay: null });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly when options box not open', () => {
    const { asFragment } = createView();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when options box open and passes all props to the child component', async () => {
    const { asFragment } = createView();

    await user.click(
      screen.getByLabelText(`More options for ${props.channel.name}`)
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('closes options box when you click outside of it', async () => {
    render(
      <div>
        <div>Outside</div>
        <MoreOptionsToggle {...props} />
      </div>
    );

    await user.click(
      screen.getByLabelText(`More options for ${props.channel.name}`)
    );

    await user.click(screen.getByText('Outside'));

    expect(screen.queryByTestId('mock-moreOptionsBox')).not.toBeInTheDocument();
  });
});
