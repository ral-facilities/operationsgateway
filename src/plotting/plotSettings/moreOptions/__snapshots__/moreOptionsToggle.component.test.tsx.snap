// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MoreOptionsToggle > renders correctly when options box not open 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-79elbk"
  >
    <button
      aria-label="More options for shotnum"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1ooguyf-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MoreVertIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`MoreOptionsToggle > renders correctly when options box open and passes all props to the child component 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-79elbk"
  >
    <button
      aria-label="More options for shotnum"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1ooguyf-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="MoreVertIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2m0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2m0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      >
        <span
          class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
    <div
      class="MuiBox-root css-2rtj2t"
      role="dialog"
    >
      <mock-moreoptionsbox
        data-testid="mock-moreOptionsBox"
      >
        channel={
  "name": "shotnum",
  "units": "mm",
  "options": {
    "visible": true,
    "colour": "colour-0",
    "lineStyle": "solid",
    "yAxis": "left"
  }
}
selectedPlotChannels=[
  {
    "name": "shotnum",
    "units": "mm",
    "options": {
      "visible": true,
      "colour": "colour-0",
      "lineStyle": "solid",
      "yAxis": "left"
    }
  },
  {
    "name": "activeArea",
    "units": "mm",
    "options": {
      "visible": true,
      "colour": "colour-1",
      "lineStyle": "solid",
      "yAxis": "left"
    }
  },
  {
    "name": "activeExperiment",
    "units": "mm",
    "options": {
      "visible": true,
      "colour": "colour-2",
      "lineStyle": "solid",
      "yAxis": "left"
    }
  }
]
changeSelectedPlotChannels=undefined

      </mock-moreoptionsbox>
    </div>
  </div>
</DocumentFragment>
`;
